src/reducers/sources.js

```diff
@@ -184,12 +191,49 @@ function updateSource(state: SourcesState, source: Object) {
   const urls = existingUrls ? [...existingUrls, source.id] : [source.id];
 
   return {
-    ...state,
+    ...updateRelativeSource(state, updatedSource),
     sources: { ...state.sources, [source.id]: updatedSource },
     urls: { ...state.urls, [source.url]: urls }
   };
 }
 
+function updateRelativeSource(
+  state: SourcesState,
+  source: Source
+): SourcesState {
+  const root = state.projectDirectoryRoot;
+  if (!underRoot(source, root)) {
+    return state;
+  }
+
+  const relativeSource = {
+    ...source,
+    relativeUrl: getRelativeUrl(source, root)
+  };
+
+  return {
+    ...state,
+    relativeSources: {
+      ...state.relativeSources,
```

Comment:
We call `updateRelativeSource` for every source. And here it seems it's doing a lot of work. Your opinion?

Reply:
oh wow - good point. I didn't think of that!

Reply:
we should definitely have a mutable helper here

Reply:
just updated. what do you think?

Reply:
Looking good :)